Allow controlling PG backend and options via init_device_mesh

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159371



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

lw committed:
Update

[ghstack-poisoned]

lw committed:
Update

[ghstack-poisoned]

wconstab reviewed (approved):
lgtm! though it would be nice to also test that the options gets passed correctly.  (or  maybe this is covered by existing tests?)

fduwjj reviewed (approved):
Thanks for the fix and leave some comments. Stamp to unblock

wanchaol commented on torch/distributed/device_mesh.py:
```diff
@@ -993,6 +1010,9 @@ def init_device_mesh(
         mesh_shape: tuple[int, ...],
         *,
         mesh_dim_names: Optional[tuple[str, ...]] = None,
+        pg_backend_and_options: Optional[
```
hmmm I would prefer to:
1. separate the pg backend and pg options, these should be separately controlled IMO
2. name this as `mesh_dim_options` instead.

lw replied:
I find it useless to prepend `mesh_dim_`: this is the function to create a device mesh, it's quite implicit!

On the other hand, I want to explicitly mention `pg_` or `group_` or something to clarify it's not the options of the mesh dim itself, but of the underlying ProcessGroup.

wanchaol replied:
> I find it useless to prepend mesh_dim_: this is the function to create a device mesh, it's quite implicit!

Sure that make sense! I was originally thinking to align it with `mesh_dim_names`, but I agree it's redundant. Shall we simply call it `options`? Maybe we should change `mesh_dim_names` to `dim_names` or simply `names` later

> On the other hand, I want to explicitly mention pg_ or group_ or something to clarify it's not the options of the mesh dim itself, but of the underlying ProcessGroup.

I wonder what does you mean by `the options of the mesh dim itself`? I thought there's nothing can be configured (or have options) on a mesh dim except its underlying communicator have options, so it would be quite explicit to user? I am ok with `group_options`, but I think if there's no confusion about `options` we should just go with `options`?

wanchaol reviewed (changes requested):
Thanks for adding the feature! I have some suggestions on how the UX should be look like, requesting changes to align on the UX.

My opinion is that we should separate the backend selection and the pg options, instead of blending them together, this would make the argument more concise. Here is the proposed UX:
```
# switch between different backends to be controlled by the device_type argument, this is aligned with how process group backend selection works
init_device_mesh("cuda:nccl", ..)
# Specify backend options can be passed in by sth like
init_device_mesh("cuda:nccl", mesh_shape, mesh_dim_names=("dp, "tp), mesh_dim_options={"dp": nccl_pg_options})
```
Basically `mesh_dim_options` should be a type of `Dict[str, C10dBackend.Options]`, I think we can support `str` as key only but adding `int` together also sounds good.

lw commented:
I'm not sure I agree with the comments on UX.
- True, this might add a burden if one wants to specify _only_ the backend type or _only_ the options, but a given options object only works for one certain backend type (e.g., `ProcessGroupNCCL.Options(...)` requires `"nccl"` as backend type), thus it doesn't seem that burdensome to ask the user to be explicit about it. Similarly, if one wants to specify only the backend type, adding `None` as a second argument doesn't seem too hard.
- However, your proposal makes it much more verbose and redundant if one wants to give _both_ a backend type _and_ an options. Currently it would be `pg_backend_and_options={"tp": ("nccl", opts)}`, with your proposal it would be `pg_backend={"tp": "nccl"}, pg_options={"tp": opts}`, thus increasing the error-proneness if the two get out-of-sync.

What I can propose is that we allow the values of the dict to also be simple `str`s or simple objects of (a subtype of) `C10dBackend.Options`, and if so we automatically turn them into a tuple by autofilling the other half with `None`.

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -1043,6 +1063,31 @@ def init_device_mesh(
 
+            if pg_backend_and_options is not None:
+                for mesh_dim_idx, mesh_dim_name in enumerate(mesh_dim_names):
+                    if mesh_dim_name in pg_backend_and_options:
+                        if mesh_dim_idx in pg_backend_and_options:
```
Either we do (like now):
```
if a:
    if b:
        raise
    do()
```
or (as you suggest)
```
if a and b:
    raise
if a:
    do()
```
I prefer the first one since it doesn't repeat `a` twice.

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -1043,6 +1063,31 @@ def init_device_mesh(
+                            pg_backend_and_options.pop(mesh_dim_name)
+                        )
+
+        if pg_backend_and_options is not None:
+            pg_backend_and_options_tuple = tuple(
+                pg_backend_and_options.pop(i, (None, None))
+                for i in range(len(mesh_shape))
+            )
+            if pg_backend_and_options:
+                raise RuntimeError(
+                    f"Found invalid keys in pg_backend_and_options: got {list(pg_backend_and_options.keys())}, "
+                    f"expected integers in range [0, {len(mesh_shape)}) or one of {mesh_dim_names or []}"
+                )
+        else:
+            pg_backend_and_options_tuple = None
```
The previous for-loop is only entered if `mesh_dim_names` is not None. So we could merge this into a single loop, but this loop would then have to do redundant checks on `mesh_dim_names`, and that would also be redundant (in a different way).

wanchaol commented:
> * True, this might add a burden if one wants to specify _only_ the backend type or _only_ the options, but a given options object only works for one certain backend type (e.g., `ProcessGroupNCCL.Options(...)` requires `"nccl"` as backend type), thus it doesn't seem that burdensome to ask the user to be explicit about it. Similarly, if one wants to specify only the backend type, adding `None` as a second argument doesn't seem too hard.

I think the problem of the blending the concept of backend and pg option is that: many times user actually only want to specify the backend, or only want to specify the options, there're very rare chance that user want to specify them together, so IMO we should try to make the common case experience be better, i.e. what if user just pass pg options, they don't want to specify the backend? when specifying options for the multi-dim mesh the backend needs to be specified again and again.

> * However, your proposal makes it much more verbose and redundant if one wants to give _both_ a backend type _and_ an options. Currently it would be `pg_backend_and_options={"tp": ("nccl", opts)}`, with your proposal it would be `pg_backend={"tp": "nccl"}, pg_options={"tp": opts}`, thus increasing the error-proneness if the two get out-of-sync.

I wonder how it make things much verbose if one want to give both a backend type and an options? My proposal is to not make a separate `pg_backend` argument, but rather encode it directly in the `device_type`, this is similar to how the processgroup backend specification done. I think it is actually simpler than the current UX, the proposed UX would be sth like this
```
init_device_mesh("cuda:nccl", mesh_shape, backend_options={"tp": opts})
```
In current way it would be:
```
init_device_mesh("cuda", mesh_shape, pg_backend_and_options={"tp": ("nccl", opts)})
```
I think it would become even simpler if user want to specify the pg options to multiple dimension of the device mesh, i.e. the proposed UX would be:
```
init_device_mesh(
    "cuda:nccl",
    mesh_shape,
    backend_options={"tp": tp_opts, "dp": dp_opts, "cp": cp_opts}
)
```
In the current way it would be:
```
init_device_mesh(
    "cuda",
    mesh_shape,
    pg_backend_and_options={"tp": ("nccl", tp_opts), "dp": ("nccl", dp_opts), "cp": ("nccl", cp_opts)}
)
```
It seems to me that repeating the "nccl" backend and with nested dict looks cumbersome. There should not be out of sync issue as the backend is parsed from device_type and we can easily guard it.

wconstab commented:
here's a clarifying question:

how important is it to support a different backend type _per_ dimension?  if so, then this version is limiting
`init_device_mesh("cuda:nccl",` because it implies one backend type shared by all the mesh dims.  I'm not sure if @lw is wanting this but i think its getting more common (e.g. torchft) to want a special type of backend for just one dimension.

wanchaol commented:
> how important is it to support a different backend type _per_ dimension? if so, then this version is limiting `init_device_mesh("cuda:nccl",` because it implies one backend type shared by all the mesh dims. I'm not sure if @lw is wanting this but i think its getting more common (e.g. torchft) to want a special type of backend for just one dimension.

@wconstab I am not sure about how torchft exactly wants to change the backend for each dimension, or I think it's not very clear to me that even having different backend on each dimension make sense, i.e. the way to create subgroup is to inherit the backend of the world group (and need to rely on the backend way to split the group, i.e, ncclSplit), and many operations of device mesh today (`flatten`, the upcoming `split`) requires to have the same backend, otherwise the operation are very much undefined.

wconstab commented:
I think we better not assume the world group is materialized. It may be important to have it for some jobs, but it is also too expensive to create a world in larger scale jobs. So I prefer the direction of decoupling the top level world from the first layer of dims.  And in this case I think it can make sense to allow separate dims to use different backends. 

But I also agree the semantic of split should have a defined meaning corresponding to a pg split.

lw commented:
- There are use cases where different dims should use different backends (e.g., intra- and inter-node)
- The root PG might not be "materialized", but we must count on it having a backend type and some options that we can copy

I'm gonna make a call in order to close this discussion, which is to proceed as follow: I'll rename the argument to `pg_override` and make it accept either a (type, opts) tuple, but _also_ just the type (str) or the options (C10dBackend.Options). I believe this should make everyone happy.

lw committed:
Update

[ghstack-poisoned]

lw commented on torch/csrc/distributed/c10d/FakeProcessGroup.hpp:
```diff
@@ -20,7 +20,21 @@ class FakeWork : public Work {
 class FakeProcessGroup : public Backend {
  public:
-  FakeProcessGroup(int rank, int size) : Backend(rank, size) {}
+  struct Options : Backend::Options {
+    explicit Options() : Backend::Options("fake") {}
+
+    int fake_option = 0;
+  };
```
Needed to define options for the fake PG in order to test the options-only override

wconstab replied:
i guess this is fine. Maybe later we can replace it with a real option

lw commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -355,22 +355,26 @@ def backend(self) -> str:
 
-    def init_pg(self, eager_init) -> None:
-        if "nccl" in self.backend and torch.cuda.device_count() < self.world_size:
+    def init_pg(self, eager_init, backend: Optional[str] = None) -> None:
```
Had to modify this (and below in this file) in order to allow using the "fake" backend as the root PG

lw commented on torch/testing/_internal/distributed/fake_pg.py:
```diff
@@ -11,7 +11,7 @@ class FakeStore(dist.Store):
 
 
-def _create_fake_pg(prefix_store, rank, world_size, timeout):
+def _create_fake_pg(common_opts, backend_opts):
```
Switched to using the "extended API" as otherwise the options wouldn't get passed to the FakePG constructor

fduwjj replied:
I am wondering it this will cause any bc breaking for some tests?

lw committed:
Update

[ghstack-poisoned]

lw commented:
I've added support for the pg override to `_flatten` as well

lw commented:
> what if user create a 4D or 5D device mesh, and they want to use their custom backend (i.e. even the fake backend testing here), but they don't need to customize the options for device mesh dimensions

I don't understand if this is a use-case you're hitting in practice or if you're imagining corner cases where the API would be awkward to use? The new API addresses what you asked earlier (one can provide just the options if one so chooses), but now you're evoking a case where one _does_ want to change the backend?

In order to pragmatically address your example use-case, I think the user should just use `fake` as their primary backend when calling `init_process_group`, and then they wouldn't have to pass any override. If for whatever reason they want the device mesh to use a different root/default backend than c10d, we could envision adding a new optional `backend=...` argument to `init_device_mesh`, which acts as a default for all dimensions. I suggest we only do that once we have a compelling use-case.

Note also that no API in PyTorch is truly private. No matter whether there's a leading underscore, if that's the only way to achieve what people need, they _will_ use it in their downstream project and PyTorch won't be able to remove it or break it. Case in point: the `_flatten` method is supposedly private but we're using it in our codebase and we'd be quite annoyed if it changed.

Do you have a concrete proposal that addresses your concerns? If not, I'll probably merge this PR as-is.

wanchaol commented:
> I don't understand if this is a use-case you're hitting in practice or if you're imagining corner cases where the API would be awkward to use? The new API addresses what you asked earlier (one can provide just the options if one so chooses), but now you're evoking a case where one _does_ want to change the backend?

@lw To be clear, my [original comment](https://github.com/pytorch/pytorch/pull/159371#issuecomment-3137495884) is about both: "user actually only want to specify the backend, or only want to specify the options". The new API only address one of the ask, so I am not sure why you think I evoke a new case, as I am just discussing the existing ones. The case that user only specify the backend is pretty common, i.e. for HCCL users, or any users who implemented custom process group and register as a backend (we have one so it's something we would hit with the current API).
  
> Note also that no API in PyTorch is truly private. No matter whether there's a leading underscore, if that's the only way to achieve what people need, they _will_ use it in their downstream project and PyTorch won't be able to remove it or break it. Case in point: the `_flatten` method is supposedly private but we're using it in our codebase and we'd be quite annoyed if it changed.

The private API is mostly for developers to have the freedom of adapt/change the API depending on new usecases/feedbacks before making it stable, otherwise every BC breaking change need to go through 3 release cycle. It's also a signal to users that PyTorch can remove/break the API so use at risk. If we don't have private APIs then we don't need to have prototype features, everything can be public and develop/use it prod, I don't think PyTorch want to be in that way. 
 
> Do you have a concrete proposal that addresses your concerns? If not, I'll probably merge this PR as-is.

I already gave my suggestion in my previous [comments](https://github.com/pytorch/pytorch/pull/159371#issuecomment-3137495884), basically we can embed the backend as a substring of the device_type, i.e. `cuda:nccl` or `cuda:fake`, this aligns with the ProcessGroup dispatchable backend API style. When that's specified we would by default initialize all process group with that backend, and if the `pg_override` override one dimension's backend, only that dimension would be overridden but still all other dimensions would still use the global backend user specified, instead of the default one.

You can also of course merge the PR as is and ignore my comments, as Meta engineers have privilege over PyTorch.

wconstab commented:
I think we could add a separate option like 'default_backend' to specify the type to use for all mesh dims.

Then we can use the new per-dimension thing (any of (backend, opts) or backend or opts) to override this default.

Personally I'm not a big fan of the old way of specifying via colon separated string. I'm also not a big fan of having more than one device type per process group. I'm not sure if device mesh currently supports this aspect of ProcessGroup- (cpu:gloo, cuda:nccl all inside one device mesh).  I'd rather keep things simpler and just have 1) device arg, 2) default_backend optional kwarg, 3) per dim overrides optional kwarg.

Would that make everyone happy?

wanchaol commented:
I think this is a very important change to DeviceMesh, so I appreciate @lw’s contribution on it. My intention of the UX discussion here is to get the best API out for PyTorch users, I think the PyTorch frontend API design is very important and it’s the main reason why PyTorch won over users. I apologize if the discussions make you feel I’m too nitpicking on APIs, etc.

I want to address the UX to allow user specifying either backend and options in a simple way. I think the current API already solved one part. But to solve both parts coherently, we need to design them together as their behaviors are tightly coupled, mainly to avoid breaking the behaviors later (i.e. change arg names to align them). This is the main reason that I want to make it a private API first as it seems the PR does not want to solve the other part immediately. I agree we do not have to implement them together. 

@wconstab I'm not very clear why you think the currently dispatchable backend API style is not good, but I am ok with not following the PG way. This does make me feel many parts of ProcessGroup designs seems not well thought (i.e. global map, dispatchable API). This is fine as things’ve evolved, but I don’t want this to happen for the DeviceMesh abstraction again. I agree with having a `backend` kwarg to allow user to specify the backend for all dimensions of device_mesh, but I feel the `pg_override` name is not very aligned with the name of `backend` (also to consider other potential communicator impls not called PG). I wonder if we can align the names to be coherent as following if folks agrees:

* `backend: str | None`: specify the backend for the device_type, if not specified, use the default backend
* `backend_options`: current PR behavior, allow user to either specify backend, or options, or both for one or multiple mesh dimensions

I'm happy to stamp the PR once we converge on the names and have plan to implement the `backend` kwarg in follow up PRs.

ezyang commented:
Just to summarize, it sounds like concretely @wanchaol is proposing that we rename `pg_override` to `backend_options`, with no other changes. Is that right?

lw commented:
I'm ok with replacing `pg_` with `backend_`. I'm less ok with replacing `_override` with `_options`, because (1) it's not just options, it can also be the backend class itself, and (2) because `_options` only tells you what kind of object it accepts (and this is already clear from the type annotation), but it doesn't tell you how it's going to be used, whereas `_override` is more self-descriptive about it.

Can everyone live with this being called `backend_override`?

lw commented:
Also, I'm totally ok with adding a catch-all `backend` argument to `init_device_mesh`, and if anyone wants to pick it up please go ahead! I don't have a need for such a feature hence I won't be pursuing it myself.

wanchaol commented:
> I'm ok with replacing `pg_` with `backend_`. I'm less ok with replacing `_override` with `_options`, because (1) it's not just options, it can also be the backend class itself, and (2) because `_options` only tells you what kind of object it accepts (and this is already clear from the type annotation), but it doesn't tell you how it's going to be used, whereas `_override` is more self-descriptive about it.
> 
> Can everyone live with this being called `backend_override`?

Yes I think the reasonings of `backend_override` make sense to me, let's go with `backend_override` then, I'll stamp to unblock.

lw committed:
Update

[ghstack-poisoned]
